File: frontend/src/hooks/questions.js
/**
 * Client-side question registry and scoring.
 * Mirrors backend app/core/questions.py and app/services/scoring.py exactly.
 * Used for instant UI updates without waiting for API round-trips.
 */

export const QUESTIONS = [
  { id:"AI-Q1",  panel:"user",   field:"systemName",          label:"AI System Name",          mandatory:false, weight:0.03,
    q:"What is the name of your AI system?",
    suggestions:["Provide a concise name that clearly reflects the system's domain and purpose"],
    hints:["Clear descriptive name","Reflects domain/purpose"] },
  { id:"AI-Q2",  panel:"user",   field:"ownerName",           label:"Business Owner",          mandatory:false, weight:0.03,
    q:"Who is the business owner responsible for this AI system?",
    suggestions:["Identify the business owner by full name, their business unit, department, and the team involved"],
    hints:["Full name","Business unit","Department","Team"] },
  { id:"AI-Q3",  panel:"user",   field:"contactEmail",        label:"Contact Email",           mandatory:false, weight:0.02,
    q:"What is the primary contact email for this submission?",
    suggestions:["Provide a valid business email address as the point of contact for follow-ups"],
    hints:["Valid business email"] },
  { id:"AI-Q4",  panel:"system", field:"problemStatement",    label:"Problem Statement",       mandatory:false, weight:0.06,
    q:"What problem or business opportunity does your AI system address?",
    suggestions:["Provide a concise 2-3 sentence description of the AI solution covering what it does, its intended business purpose, target users, and whether it is new or enhances existing capability","Identify the business process or workflow that the AI solution supports"],
    hints:["Articulate the pain point","Quantify business impact","Who is affected","Current vs desired state"] },
  { id:"AI-Q5",  panel:"system", field:"aiSolution",          label:"AI Solution",             mandatory:false, weight:0.06,
    q:"Describe your AI solution — how it works and what it does.",
    suggestions:["Provide a concise 2-3 sentence description of the AI solution covering what it does, its intended business purpose, target users, and whether it is new or enhances existing capability","Explain how the AI solution integrates with existing systems"],
    hints:["Plain-language description","Core AI/ML capabilities","Inputs and outputs"] },
  { id:"AI-Q6",  panel:"system", field:"valueProposition",    label:"Value Proposition",       mandatory:true,  weight:0.12,
    q:"What is the expected business value and impact of this AI system?",
    suggestions:["Provide a concise 2-3 sentence description of the AI solution covering what it does, its intended business purpose, target users, and whether it is new or enhances existing capability","Specify the deployment model (SaaS vendor-hosted, self-hosted, or hybrid) and indicate if alternative hosting options are available"],
    hints:["Quantify ROI or cost savings","Impact on users","Short and long-term benefits","Specific KPIs"] },
  { id:"AI-Q7",  panel:"system", field:"targetUsers",         label:"Target Users",            mandatory:true,  weight:0.08,
    q:"Who are the primary users or beneficiaries of this AI system?",
    suggestions:["Provide a concise 2-3 sentence description of the AI solution covering what it does, its intended business purpose, target users, and whether it is new or enhances existing capability","Identify the key stakeholders for the AI solution"],
    hints:["Internal vs external users","User roles","Estimated number affected","Key stakeholders"] },
  { id:"AI-Q13", panel:"system", field:"explainability",      label:"Explainability",          mandatory:false, weight:0.04,
    q:"How explainable is your AI system's decision-making?",
    suggestions:["Describe how decisions made by the AI are explained to end users","Identify any explainability tools used (e.g. SHAP, LIME)"],
    hints:["How decisions are explained","Explainability tools","User-facing explanations"] },
  { id:"AI-Q14", panel:"system", field:"humanOversight",      label:"Human Oversight",         mandatory:false, weight:0.04,
    q:"What human oversight and control mechanisms are in place?",
    suggestions:["Describe human-in-the-loop processes, override capabilities, escalation paths, and monitoring"],
    hints:["Human-in-the-loop","Override capabilities","Escalation paths"] },
  { id:"AI-Q15", panel:"system", field:"usesPersonalData",    label:"Uses Personal Data",      mandatory:false, weight:0.02,
    q:"Does this AI system use, process, or generate personal data? (Yes / No)",
    suggestions:["Confirm whether the use case processes any Personal Information (PI), Confidential Information (CI), or only uses public/non-sensitive data (green/yellow classification)"],
    hints:["Yes or No","Personal data includes names, emails, IDs, location, behaviour"],
    isConditionalTrigger:true },
  { id:"AI-Q16", panel:"system", field:"riskAssessment",      label:"Risk Assessment",         mandatory:false, weight:0.05,
    q:"What are the key risks associated with this AI system?",
    suggestions:["Identify model failure modes, bias risks, adversarial attack vectors, and regulatory/reputational risks"],
    hints:["Model failure modes","Bias risks","Regulatory risks"] },
  { id:"AI-Q17", panel:"system", field:"riskMitigation",      label:"Risk Mitigation",         mandatory:true,  weight:0.10,
    q:"What specific controls and mitigations manage AI risks?",
    suggestions:["Describe bias detection and mitigation measures","Explain model drift monitoring and retraining approach","Detail fallback mechanisms and human review triggers"],
    hints:["Bias detection","Model drift monitoring","Fallback mechanisms","Human review triggers"] },
  { id:"AI-Q18", panel:"system", field:"testingValidation",   label:"Testing & Validation",    mandatory:false, weight:0.04,
    q:"How has this AI system been tested and validated?",
    suggestions:["Describe evaluation datasets, fairness and bias testing, performance metrics, and user acceptance testing"],
    hints:["Evaluation datasets","Fairness testing","Performance metrics","UAT"] },
  { id:"AI-Q20", panel:"system", field:"personalDataDetails", label:"Personal Data Details",   mandatory:true,  weight:0.06,
    q:"Describe what personal data is processed and the legal basis.",
    suggestions:["Confirm the data classification level and whether the use case processes any Personal Information (PI), Confidential Information (CI), or only uses public/non-sensitive data (green/yellow classification)"],
    hints:["Categories of personal data","Legal basis (GDPR)","Data minimization","Retention and deletion"],
    conditionalOn:{qid:"AI-Q15",kws:["yes"]} },
  { id:"AI-Q8",  panel:"tech",   field:"technicalApproach",   label:"Technical Approach",      mandatory:false, weight:0.05,
    q:"What AI/ML techniques, models, or frameworks does your system use?",
    suggestions:["Identify the data sources used by the AI solution","Provide a concise description of the AI/ML models, algorithms, and technology stack deployed"],
    hints:["Models or algorithms","Training data","Technology stack","Deployment"] },
  { id:"AI-Q9",  panel:"tech",   field:"dataDescription",     label:"Data Description",        mandatory:false, weight:0.05,
    q:"What data does this AI system use? Describe sources, types, and volumes.",
    suggestions:["Identify the data sources used by the AI solution","Describe data types (structured/unstructured), volume, and update frequency"],
    hints:["All data sources","Data types","Volume and update frequency","Data quality"] },
  { id:"AI-Q10", panel:"tech",   field:"integrations",        label:"Integrations",            mandatory:false, weight:0.03,
    q:"What systems or APIs does this AI system integrate with?",
    suggestions:["Explain how the AI solution integrates with existing systems","Identify upstream and downstream system dependencies and integration methods"],
    hints:["Upstream and downstream dependencies","Integration method","Cloud platforms"] },
  { id:"AI-Q11", panel:"tech",   field:"scalability",         label:"Scalability",             mandatory:false, weight:0.03,
    q:"How does your system handle scale and performance requirements?",
    suggestions:["Describe the expected load, scaling approach, latency, throughput, and failover strategy"],
    hints:["Expected load","Scaling approach","Latency and throughput","Failover"] },
  { id:"AI-Q12", panel:"tech",   field:"dataPrivacy",         label:"Data Privacy & Security", mandatory:true,  weight:0.10,
    q:"How does your AI system handle data privacy, security, and compliance?",
    suggestions:["Provide a concise 2-3 sentence description of the AI solution covering what it does, its intended business purpose, target users, and whether it is new or enhances existing capability","Identify the regulatory requirements applicable to the AI solution"],
    hints:["PII handling and anonymization","Encryption","Regulatory compliance (GDPR, HIPAA)","Access controls","Audit logging"] },
  { id:"AI-Q19", panel:"tech",   field:"deploymentPlan",      label:"Deployment Plan",         mandatory:false, weight:0.03,
    q:"What is your deployment plan and rollout strategy?",
    suggestions:["Describe the phased rollout approach, pilot testing plan, rollback strategy, and expected go-live timeline"],
    hints:["Phased rollout","Pilot testing","Rollback plan","Go-live timeline"] },
  { id:"AI-Q22", panel:"tech",   field:"dataAuditApproach",   label:"Data Audit Approach",     mandatory:true,  weight:0.04,
    q:"How do you audit data usage? All data or sampling? (No / Yes - all data / Yes - a sampling)",
    suggestions:["Verify whether any AI functionality impacts patient safety, product quality, or assists in manufacturing/testing of products","Verify whether the use case has any impact on patients or customers"],
    hints:["Audit logging approach","Events and accesses tracked","Audit retention period"],
    conditionalOn:{qid:"AI-Q15",kws:["yes"]}, isConditionalTrigger:true },
  { id:"AI-Q23", panel:"tech",   field:"samplingRationale",   label:"Sampling Rationale",      mandatory:true,  weight:0.05,
    q:"Why is full data audit not performed? Justify the approach.",
    suggestions:["Verify whether any AI functionality impacts patient safety, product quality, or assists in manufacturing/testing","Verify whether the use case has any impact on patients or customers"],
    hints:["Technical or business constraints","Sampling methodology","Risk mitigation","Compensating controls"],
    conditionalOn:{qid:"AI-Q22",kws:["no","yes - a sampling"]} },
];

export const Q_MAP = Object.fromEntries(QUESTIONS.map(q => [q.id, q]));

export const PANELS = {
  user:   { label:"User Info",   accent:"#7c8fff", icon:"◎", desc:"Ownership and contact details." },
  system: { label:"System Info", accent:"#00e5a0", icon:"⬡", desc:"Business context, value, governance and risk." },
  tech:   { label:"Tech Info",   accent:"#ff9d4d", icon:"◈", desc:"Architecture, data, privacy and deployment." },
};

// Mirrors get_active_question_ids() from backend exactly
export function getActiveQIds(rawAnswers) {
  const active = new Set(QUESTIONS.map(q => q.id));
  const txt = qid => String(rawAnswers[qid] ?? "").toLowerCase();
  const has = (qid, kws) => kws.some(k => txt(qid).includes(k.toLowerCase()));

  const ai15done = "AI-Q15" in rawAnswers;
  const ai15yes  = has("AI-Q15", ["yes"]);

  if (!ai15done || !ai15yes) {
    active.delete("AI-Q20"); active.delete("AI-Q22"); active.delete("AI-Q23");
  } else if ("AI-Q22" in rawAnswers && !has("AI-Q22", ["no","yes - a sampling"])) {
    active.delete("AI-Q23");
  }
  return active;
}

// Mirrors aggregate_approval() from backend exactly
const MAND_THRESHOLD = 0.6, STATIC_PENALTY = 0.15, DEFAULT_MAND_C = 0.3;

export function aggregateApproval(confScores, activeQIds) {
  const mandatory = new Set(QUESTIONS.filter(q => q.mandatory && activeQIds.has(q.id)).map(q => q.id));
  const normC = {}; const inc = []; let totW = 0;
  for (const q of QUESTIONS) {
    if (!activeQIds.has(q.id)) continue;
    const raw = confScores[q.id];
    const c = raw == null ? (mandatory.has(q.id) ? DEFAULT_MAND_C : null) : Math.min(1, Math.max(0, raw));
    if (c == null) continue;
    normC[q.id] = c;
    if (q.weight > 0) { inc.push(q.id); totW += q.weight; }
  }
  if (!inc.length || !totW) return 0;
  const base = inc.reduce((s, id) => s + (Q_MAP[id].weight / totW) * normC[id], 0);
  let pen = 1.0;
  for (const id of mandatory)
    if ((normC[id] ?? DEFAULT_MAND_C) < MAND_THRESHOLD) pen *= (1 - STATIC_PENALTY);
  return Math.min(1, base * pen);
}

export function panelConf(panel, confScores, activeQIds) {
  const qs = QUESTIONS.filter(q => q.panel === panel && activeQIds.has(q.id) && confScores[q.id] != null);
  if (!qs.length) return null;
  const totW = qs.reduce((s, q) => s + q.weight, 0);
  return totW ? qs.reduce((s, q) => s + (q.weight / totW) * confScores[q.id], 0) : null;
}

export const scoreColor = s => s==null?"#1e2e44":s>=0.8?"#00e5a0":s>=0.6?"#7c8fff":s>=0.4?"#f5c542":"#ff4d6d";
export const scoreLabel = s => s==null?"—":s>=0.8?"Strong":s>=0.6?"Good":s>=0.4?"Fair":"Weak";
export const scorePct   = s => s==null?"—":`${Math.round(s*100)}%`;

File: backend/app/core/questions.py
"""
SAGE Questions Registry — single source of truth for all AI governance questions.
Maps to the AI Review submission form. Suggestions[] drive scoring + agent follow-ups.
"""
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class ConditionalOn:
    qid: str
    kws: List[str]


@dataclass
class Question:
    id: str
    panel: str               # "system" | "user" | "tech"
    field: str               # camelCase field name
    label: str
    mandatory: bool
    weight: float
    q: str                   # Question text shown to user
    suggestions: List[str]   # Drives scoring
    hints: List[str]
    is_conditional_trigger: bool = False
    conditional_on: Optional[ConditionalOn] = None


QUESTIONS: List[Question] = [
    # ── SYSTEM INFORMATION PANEL ──────────────────────────────────────────────
    Question("AI-Q2", "system", "systemTitle", "System Title", True, 0.05,
        "What is the title of the System?",
        ["Provide a clear and descriptive title for the AI system"],
        ["Clear system name", "Reflects purpose"]),

    Question("AI-Q3", "system", "solutionOverview", "Solution Overview", True, 0.08,
        "Provide an overview of the solution.",
        ["Provide a concise description of what the system does, its purpose, and how it works",
         "Describe the key capabilities and functions of the solution"],
        ["What the system does", "How it works", "Key capabilities"]),

    Question("AI-Q4", "system", "valueProposition", "Value Proposition", True, 0.07,
        "What is the value proposition for this solution?",
        ["Describe the business value, benefits, and impact this solution provides",
         "Explain the problem it solves and the value it delivers to the organization"],
        ["Business value", "Benefits", "Problem solved", "Impact"]),

    Question("AI-Q5", "system", "sponsorOrg", "Sponsor Organization", True, 0.04,
        "Which organization will sponsor the System?",
        ["Identify the sponsoring organization or business unit responsible for the system"],
        ["Organization name", "Business unit"]),

    Question("AI-Q6", "system", "vendorProduct", "Vendor Product", False, 0.02,
        "Are you planning to implement a vendor product as part or all of this system? (Yes / No)",
        ["Indicate whether a vendor product will be used as part of or for the entire system"],
        ["Yes or No"],
        is_conditional_trigger=True),

    Question("AI-Q7", "system", "vendorNames", "Vendor Names", False, 0.03,
        "Which vendor(s) are you planning to use?",
        ["List all vendor names that will supply products or services for this system"],
        ["Vendor names"],
        conditional_on=ConditionalOn("AI-Q6", ["yes"])),

    Question("AI-Q8", "system", "productNames", "Product Names", False, 0.03,
        "What product(s) from the vendor(s) will you use?",
        ["List the specific product names from each vendor"],
        ["Product names", "Product versions if known"],
        conditional_on=ConditionalOn("AI-Q6", ["yes"])),

    # ── USER INFORMATION PANEL ────────────────────────────────────────────────
    Question("AI-Q9", "user", "userDescription", "User Description", True, 0.08,
        "Briefly describe the users of the system, what they will be using it for, and what problem it will solve for them.",
        ["Describe who the users are, their roles, what they will use the system for, and the problem it solves for them",
         "Identify the target user groups and their needs"],
        ["Who are the users", "What they use it for", "Problem it solves"]),

    Question("AI-Q10", "user", "deploymentLocation", "Deployment Location", True, 0.04,
        "Where will you deploy your System? (Select all that apply)",
        ["Identify all deployment locations and environments for the system (e.g., cloud, on-premises, hybrid, specific regions)"],
        ["Cloud", "On-premises", "Hybrid", "Specific regions"]),

    Question("AI-Q11", "user", "audience", "System Audience", True, 0.04,
        "Who is the audience of this system? (Select all that apply)",
        ["Identify all audiences for the system (e.g., internal employees, external customers, partners, patients, HCPs)"],
        ["Internal employees", "External customers", "Partners", "Other audiences"]),

    # ── TECHNICAL INFORMATION PANEL ───────────────────────────────────────────
    Question("AI-Q12", "tech", "additionalTechInfo", "Additional Technical Info", False, 0.02,
        "Do you have any additional technical information to provide? (Yes / No)",
        ["Indicate whether additional technical details are available at this time"],
        ["Yes or No"],
        is_conditional_trigger=True),

    Question("AI-Q13", "tech", "systemMaturity", "System Maturity", True, 0.04,
        "What is the maturity of the System?",
        ["Describe the maturity level of the system (e.g., Concept/Ideation, Development, Pilot, Production, Retired)"],
        ["Concept/Ideation", "Development", "Pilot", "Production"]),

    Question("AI-Q14", "tech", "dataProcessed", "Data Processed", True, 0.06,
        "Does the System process any of the following data? (Select all that apply: Personal Information, Confidential Information, Patient Data, Employee Data, etc.)",
        ["Identify all types of data the system processes including Personal Information (PI), Confidential Information (CI), patient data, employee data, or other sensitive data",
         "If the AI use case processes Personal Information, a Privacy Review is required per the global Respecting Privacy procedure"],
        ["Personal Information", "Confidential Information", "Patient data", "Employee data", "Public data"]),

    Question("AI-Q15", "tech", "dataClassification", "Data Classification", True, 0.05,
        "What is the highest data classification for processed data? Note: If data classification increases during implementation, a new review cycle will be required.",
        ["Identify the highest data classification level (e.g., Public, Internal, Confidential, Restricted)"],
        ["Public", "Internal", "Confidential", "Restricted"]),

    Question("AI-Q16", "tech", "dataForTraining", "Data Used for Training", False, 0.04,
        "Is any of this data used for training the model? (Yes / No / Not Sure)",
        ["Indicate whether any of the processed data is used for training the AI/ML model"],
        ["Yes", "No", "Not Sure"]),

    Question("AI-Q17", "tech", "dataNeeded", "Data Needed", True, 0.05,
        "Describe the Data Needed.",
        ["Describe all data sources, types, and volumes needed by the system",
         "Explain how data flows into and out of the system"],
        ["Data sources", "Data types", "Data volumes", "Data flow"]),

    Question("AI-Q18", "tech", "aiFunctionality", "AI Functionality", True, 0.06,
        "Does the AI Functionality of the system do any of the following? (Select all that apply: generate content, make predictions, classify data, recommend actions, automate decisions, etc.)",
        ["Describe all AI functionalities of the system such as content generation, predictions, classification, recommendations, decision automation, or other AI capabilities"],
        ["Content generation", "Predictions", "Classification", "Recommendations", "Decision automation"]),

    Question("AI-Q19", "tech", "humanReview", "Human Review of Output", True, 0.05,
        "Is the AI System output reviewed by a human before it is used? (Yes – 100% of the output / Yes – a sampling / No)",
        ["Describe the extent of human review of AI system outputs before they are used",
         "If sampling, describe the sampling methodology"],
        ["Yes – 100%", "Yes – a sampling", "No"],
        is_conditional_trigger=True),

    Question("AI-Q20", "tech", "aiPlatforms", "AI Platforms & Technologies", True, 0.05,
        "Which AI platforms/technologies will be used? Include AI platforms/technology, Models, and Open Source Models.",
        ["List all AI platforms, technologies, models (proprietary and open source) that will be used",
         "Specify model names and versions where known"],
        ["AI platforms", "Proprietary models", "Open source models", "Technology stack"]),

    Question("AI-Q21", "tech", "continuousLearning", "Continuous Learning", False, 0.03,
        "Does the AI/ML use continuous learning? (Yes / No)",
        ["Indicate whether the AI/ML system uses continuous learning (model updates automatically from new data)"],
        ["Yes or No", "How the model is updated"]),

    Question("AI-Q22", "tech", "privacyRequestId", "Privacy Request Identifier", False, 0.02,
        "What is the Privacy Request Identifier?",
        ["Provide the Privacy Request Identifier if a privacy review has been initiated"],
        ["Privacy Request ID"]),

    Question("AI-Q23", "tech", "saeReviewNumber", "SAE Review Number", False, 0.02,
        "What is the SAE review number?",
        ["Provide the SAE (Security Architecture Evaluation) review number if available"],
        ["SAE review number"]),

    Question("AI-Q24", "tech", "wwtpReview", "WwTP Review", False, 0.02,
        "What is the WwTP review status or number?",
        ["Provide the WwTP (Ways We Trust People) review number or status if available"],
        ["WwTP review number or status"]),
]

Q_MAP: dict[str, Question] = {q.id: q for q in QUESTIONS}

PANELS = {
    "system": {"label": "System Info",    "accent": "#00e5a0", "icon": "⬡", "desc": "System details, value proposition, and vendor information."},
    "user":   {"label": "User Info",      "accent": "#7c8fff", "icon": "◎", "desc": "Users, deployment, and audience."},
    "tech":   {"label": "Technical Info",  "accent": "#ff9d4d", "icon": "◈", "desc": "Technical details, data, AI functionality, and compliance."},
}


def get_active_question_ids(raw_answers: dict) -> set:
    """
    Returns active question IDs based on conditional logic.
    raw_answers must be keyed by question ID e.g. {"AI-Q6": "Yes"}.
    Exactly mirrors getActiveQIds() from the React frontend.
    """
    active = {q.id for q in QUESTIONS}

    def txt(qid: str) -> str:
        return str(raw_answers.get(qid, "")).lower()

    def has_kw(qid: str, kws: list) -> bool:
        return any(k.lower() in txt(qid) for k in kws)

    # AI-Q6 (Vendor Product) → AI-Q7 (Vendor Names), AI-Q8 (Product Names)
    ai6_done = "AI-Q6" in raw_answers
    ai6_yes  = has_kw("AI-Q6", ["yes"])
    if not ai6_done or not ai6_yes:
        active.discard("AI-Q7")
        active.discard("AI-Q8")

    return active
